tap/tapify.py

```diff
@@ -1,106 +1,302 @@
+        args_data.append(arg_data)
+    return _TapData(args_data, has_kwargs, known_only)
+
+
+def _tap_data_from_class_or_function(
+    class_or_function: _ClassOrFunction, func_kwargs: dict[str, Any], param_to_description: dict[str, str]
+) -> _TapData:
+    """
+    Note
+    ----
+    Deletes redundant keys from `func_kwargs`
+    """
+    args_data: List[_ArgData] = []
     has_kwargs = False
+    known_only = False
 
-    # Add arguments of class init or function to the Tap object
-    for param_name, param in sig.parameters.items():
-        tap_kwargs = {}
+    sig = inspect.signature(class_or_function)
 
+    for param_name, param in sig.parameters.items():
         # Skip **kwargs
-        if param.kind == Parameter.VAR_KEYWORD:
+        if param.kind == inspect.Parameter.VAR_KEYWORD:
             has_kwargs = True
             known_only = True
             continue
 
-        # Get type of the argument
-        if param.annotation != Parameter.empty:
-            # Any type defaults to str (needed for dataclasses where all non-default attributes must have a type)
-            if param.annotation is Any:
-                tap._annotations[param.name] = str
-            # Otherwise, get the type of the argument
-            else:
-                tap._annotations[param.name] = param.annotation
+        if param.annotation != inspect.Parameter.empty:
+            annotation = param.annotation
+        else:
+            annotation = Any
 
-        # Get the default or required of the argument
         if param.name in func_kwargs:
-            tap_kwargs["default"] = func_kwargs[param.name]
+            is_required = False
+            default = func_kwargs[param.name]
             del func_kwargs[param.name]
-        elif param.default != Parameter.empty:
-            tap_kwargs["default"] = param.default
+        elif param.default != inspect.Parameter.empty:
+            is_required = False
+            default = param.default
         else:
-            tap_kwargs["required"] = True
+            is_required = True
+            default = inspect.Parameter.empty  # Can be set to anything. It'll be ignored
+
+        arg_data = _ArgData(
+            name=param_name,
+            annotation=annotation,
+            is_required=is_required,
+            default=default,
+            description=param_to_description.get(param.name),
+        )
+        args_data.append(arg_data)
+    return _TapData(args_data, has_kwargs, known_only)
+
+
+def _is_data_model(obj: Any) -> bool:
+    return dataclasses.is_dataclass(obj) or _is_pydantic_base_model(obj)
+
+
+def _docstring(class_or_function) -> Docstring:
+    """Parse class or function docstring in one line"""
+    is_function = not inspect.isclass(class_or_function)
+    if is_function or _is_pydantic_base_model(class_or_function):
+        doc = class_or_function.__doc__
+    else:
+        doc = class_or_function.__init__.__doc__ or class_or_function.__doc__
```

Comment:
The `or ...` part is new. IME, it's [more standard](https://realpython.com/documenting-python-code/#class-docstrings) to put a class' docstring right under the `class ...:` definition. The `__init__.__doc__` has priority to keep this PR backwards-compatible

Reply:
Hmm I'm thinking we should change the `else` to

```python
doc = class_or_function.__doc__ or class_or_function.__init__.__doc__
```

or

```python
doc = (class_or_function.__doc__ or "") + (class_or_function.__init__.__doc__ or "")
```

For a class which doesn't implement `__init__` (often the case with subclasses), `class_or_function.__init__.__doc__` defaults to this uninformative docstring:

```python
'Initialize self.  See help(type(self)) for accurate signature.'
```

If `class_or_function.__doc__` is used, then #129 can be closed by asking users to supply parameter descriptions in the class docstring.